src/Coordination/KeeperStateManager.cpp

```diff
@@ -261,10 +266,104 @@ void KeeperStateManager::save_config(const nuraft::cluster_config & config)
+    }();
+
+    return old_path;
+}
+
 void KeeperStateManager::save_state(const nuraft::srv_state & state)
 {
-    nuraft::ptr<nuraft::buffer> buf = state.serialize();
-    server_state = nuraft::srv_state::deserialize(*buf);
+    const auto & old_path = getOldServerStatePath();
+
+    if (std::filesystem::exists(server_state_path))
+        std::filesystem::rename(server_state_path, old_path);
+
+    WriteBufferFromFile server_state_file(server_state_path, DBMS_DEFAULT_BUFFER_SIZE, O_TRUNC | O_CREAT | O_WRONLY);
+    auto buf = state.serialize();
+
+    // calculate checksum
+    SipHash hash;
+    hash.update(reinterpret_cast<const char *>(buf->data_begin()), buf->size());
+    writeIntBinary(hash.get64(), server_state_file);
+
+    server_state_file.write(reinterpret_cast<const char *>(buf->data_begin()), buf->size());
+    server_state_file.sync();
+    server_state_file.close();
+
+    std::filesystem::remove(old_path);
+}
+
+nuraft::ptr<nuraft::srv_state> KeeperStateManager::read_state()
+{
+    const auto & old_path = getOldServerStatePath();
+
+    const auto try_read_file = [this](const auto & path) -> nuraft::ptr<nuraft::srv_state>
+    {
+        ReadBufferFromFile read_buf(path);
+        auto content_size = read_buf.getFileSize();
+
+        if (content_size == 0)
+            return nullptr;
+
+        try
+        {
+            uint64_t read_checksum;
+            readIntBinary(read_checksum, read_buf);
+
+            auto buffer_size = content_size - sizeof read_checksum;
+            auto state_buf = nuraft::buffer::alloc(buffer_size);
+            read_buf.read(reinterpret_cast<char *>(state_buf->data_begin()), buffer_size);
+
+            SipHash hash;
+            hash.update(reinterpret_cast<const char *>(state_buf->data_begin()), state_buf->size());
+
+            if (read_checksum != hash.get64())
+            {
+                LOG_WARNING(
+                    logger,
+                    "Invalid checksum while reading state from {}. Got {}, expected {}",
+                    path.generic_string(),
+                    hash.get64(),
+                    read_checksum);
+                return nullptr;
+            }
+
+            auto state = nuraft::srv_state::deserialize(*state_buf);
+            LOG_INFO(logger, "Read state from {}", path.generic_string());
+            return state;
+        }
+        catch (...)
+        {
+            LOG_WARNING(logger, "Failed to deserialize state from {}", path.generic_string());
+            return nullptr;
+        }
+    };
+
+    if (std::filesystem::exists(server_state_path))
+    {
+        auto state = try_read_file(server_state_path);
+
+        if (state)
+            return state;
+    }
+
+    if (std::filesystem::exists(old_path))
+    {
+        auto state = try_read_file(server_state_path);
+
+        if (state)
+            return state;
+
+        std::filesystem::rename(old_path, server_state_path);
```

Comment:
The question is what we do if we notice files, but fail to deserialize all of them. We are in an invalid state, and just writing a log maybe isn't desirable from the user because the node could end up in a weird state. We could abort Keeper if we are in this state, and add config/flag to start Keeper and ignore the invalid state, making it start from term 0 (rely on the cluster to properly update the term of it).

We are talking about an edge case here, so IMO it's okay if we push it like this because the point of this is to make Keeper work properly in general. It's a more complex logic and it could slow down merging this PR so I would do it in a different PR (I don't expect it to happen because the writes are not that heavy).